File: spoj/classicals/PRUBALL.py


# def solve(n, k):
# 	if k == 0 or k == 1: return k;
# 	if n == 1:
# 		return k
# 	return min(max(solve(n-1, x-1), solve(n, k-x)) for x in xrange(1,k+1)) + 1


def solve(n, k):
	dp = [[0 for i in xrange(k+1)] for j in xrange(n+1)]
	for i in xrange(n+1):
		for j in xrange(k+1):
			if j == 0 or j == 1 or i == 1:
				dp[i][j] = j
			else:
				dp[i][j] = min(max(dp[i-1][x-1], dp[i][j-x]) for x in xrange(1,j+1))+1
	return dp[n][k]


n = input()
k = input()
print solve(n, k)

File: spoj/classicals/archive/PRUBALL.cpp
#include <cstdio>
using namespace std;
int max(int a, int b){
	if (a > b) return a;
	return b;
}

int solve(int n, int k){
	int dp[n+1][k+1];
	for (int i = 0; i <= n; i += 1){
		for (int j = 0; j <= k; j += 1){
			if (j == 0 || j == 1 || i == 1){
				dp[i][j] = j;
			}else{
				dp[i][j] = 10000000;
				for (int x = 1; x < j+1; x += 1){
					int temp = max(dp[i-1][x-1], dp[i][j-x]);
					if (temp < dp[i][j]) dp[i][j] = temp;
				}
				dp[i][j] += 1;
			}
		}
	}
	return dp[n][k];
}

int main(void){
	int t, num, n, k;
	scanf("%d", &t);
	for (int i = 1; i <= t; i += 1){
		scanf("%d %d %d", &num, &n, &k);
		printf("%d %d\n", num, solve(n, k));
	}
}